File: lib/services/executors/claude-agent/src/internal-mcp-tools.ts
// Copyright © 2026 Fall Guy Consulting.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.apache at the repo root.

import { z } from "zod";
import {
  ATTRIBUTES_TOOL_DEFINITIONS,
  AttributesReadInput,
  AttributesSetInput,
} from "./attributes-tools.js";

export const ReportCompleteInput = z.object({
  token: z.string(),
  attributes_delta: z.record(z.unknown()).optional(),
  changed: z.boolean(),
  change_summary: z.string().nullable().optional(),
  signoffs: z.array(z.string()).optional(),
});

export const ReportBlockedInput = z.object({
  token: z.string(),
  reason: z.string(),
  context: z.unknown().optional(),
});

export const ReportErrorInput = z.object({
  token: z.string(),
  error_class: z.string(),
  payload: z.unknown().optional(),
});

export const PARK_REASONS = [
  "await_callback",
  "snooze",
] as const;

export const ReportParkInput = z.object({
  token: z.string(),
  reason: z.enum(PARK_REASONS),
  reason_note: z.string().optional(),
  resume_at: z.string().optional(),
});

export { AttributesReadInput, AttributesSetInput };

export interface ToolDefinition {
  name: string;
  description: string;
  inputSchema: object;
}

export const TOOL_DEFINITIONS: ToolDefinition[] = [
  {
    name: "report_complete",
    description:
      "Report successful completion. Optional attributes_delta carries " +
      "the terminal-final attribute writeback; omit when the run used " +
      "incremental attributes_set calls.",
    inputSchema: {
      type: "object",
      required: ["token", "changed"],
      properties: {
        token: { type: "string" },
        attributes_delta: { type: "object" },
        changed: { type: "boolean" },
        change_summary: { type: ["string", "null"] },
        signoffs: {
          type: "array",
          items: { type: "string" },
          description:
            "base64 Ed25519 sign-off signatures, when the node requires sign-offs",
        },
      },
    },
  },
  {
    name: "report_blocked",
    description: "Report that work cannot continue. Treated as agent_blocked.",
    inputSchema: {
      type: "object",
      required: ["token", "reason"],
      properties: {
        token: { type: "string" },
        reason: { type: "string" },
        context: {},
      },
    },
  },
  {
    name: "report_error",
    description:
      "Report a structured failure matching the cell's declared error class.",
    inputSchema: {
      type: "object",
      required: ["token", "error_class"],
      properties: {
        token: { type: "string" },
        error_class: { type: "string" },
        payload: {},
      },
    },
  },
  {
    name: "report_park",
    description:
      "Park the dispatch. The supervisor pauses the node until resume_at " +
      "elapses or an invalidate wakes it. Per 2026-05-14 Piece 2, reason " +
      "is the typed ParkReason value (snake_case).",
    inputSchema: {
      type: "object",
      required: ["token", "reason"],
      properties: {
        token: { type: "string" },
        reason: {
          type: "string",
          enum: [...PARK_REASONS],
          description: "Typed park reason. The agent must pick one.",
        },
        reason_note: {
          type: "string",
          description: "Optional human-readable annotation.",
        },
        resume_at: {
          type: "string",
          format: "date-time",
          description: "Optional ISO 8601 timestamp at which to wake. Absent means signal-only.",
        },
      },
    },
  },
  ...ATTRIBUTES_TOOL_DEFINITIONS,
];

File: lib/services/executors/claude-agent/internalmcptools.go
// Copyright © 2026 Fall Guy Consulting.
// SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-FallGuy-Commercial

package claudeagent

const CallbackMCPServerName = "rimsky-callback"

type ToolDefinition struct {
	Name        string
	Description string
	InputSchema map[string]any
}

func toolDefinitions() []ToolDefinition {
	return []ToolDefinition{
		{
			Name: "report_complete",
			Description: "Report successful completion of this dispatch. Call exactly once at the end of the run. " +
				"`changed: true` if the work modified files; `changed: false` for no-op reports. " +
				"`attributes_delta` carries the terminal-final attribute writeback (omit when the run " +
				"made no attribute changes).",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token", "changed"},
				"properties": map[string]any{
					"token":            map[string]any{"type": "string"},
					"attributes_delta": map[string]any{"type": "object"},
					"changed":          map[string]any{"type": "boolean"},
					"change_summary":   map[string]any{"type": []any{"string", "null"}},
					"signoffs": map[string]any{
						"type":  "array",
						"items": map[string]any{"type": "string"},
						"description": "base64 Ed25519 sign-off signatures, when the node " +
							"requires sign-offs",
					},
				},
			},
		},
		{
			Name:        "report_blocked",
			Description: "Report that work cannot continue (e.g. waiting on an external signal). Treated as agent_blocked.",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token", "reason"},
				"properties": map[string]any{
					"token":   map[string]any{"type": "string"},
					"reason":  map[string]any{"type": "string"},
					"context": map[string]any{},
				},
			},
		},
		{
			Name:        "report_error",
			Description: "Report a terminal error. The dispatch is failed with the supplied error_class.",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token", "error_class"},
				"properties": map[string]any{
					"token":       map[string]any{"type": "string"},
					"error_class": map[string]any{"type": "string"},
					"payload":     map[string]any{},
				},
			},
		},
		{
			Name: "report_park",
			Description: "Park the dispatch. The supervisor pauses the node until resume_at " +
				"elapses, then re-dispatches it.",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token", "resume_at"},
				"properties": map[string]any{
					"token": map[string]any{"type": "string"},
					"resume_at": map[string]any{
						"type":        "string",
						"format":      "date-time",
						"description": "Required ISO 8601 timestamp at which to wake.",
					},
				},
			},
		},
		{
			Name: "dispatch_context_read",
			Description: "Read the dispatch's identity and disposition as captured at executor " +
				"spawn. Returns dispatch_id and run_scope_id (always present as " +
				"strings, possibly empty in non-supervised invocations), and " +
				"prior_dispatch_id + prior_dispatch_disposition (one of " +
				"stale_recovery / retry_after_error / recalculate) when this dispatch " +
				"supersedes a predecessor. Returns the same snapshot for the " +
				"duration of the run.",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token"},
				"properties": map[string]any{
					"token": map[string]any{"type": "string"},
				},
			},
		},
		{
			Name: "attributes_read",
			Description: "Read the per-run attributes object as captured at executor spawn. " +
				"Returns the same snapshot for the duration of the run.",
			InputSchema: map[string]any{
				"type":     "object",
				"required": []any{"token"},
				"properties": map[string]any{
					"token": map[string]any{"type": "string"},
				},
			},
		},
	}
}
